server/lib/manifest.py

```diff
@@ -0,0 +1,395 @@
+        :param uri: The item's URI in the manifest, typically it's path
+        :param bundle: An optional bundle that's needed for externally defined data
+        :param parent_dataset: The ID of an optional parent dataset
+        :return: Dictionary representing an aggregated file
+        """
+        aggregation = dict()
+        aggregation['uri'] = uri
+        if bundle:
+            aggregation['bundledAs'] = bundle
+        if parent_dataset:
+            aggregation['schema:isPartOf'] = parent_dataset
+        return aggregation
+
+    def add_item_records(self):
+        """
+        Creates records for a set of item ids. This is desired when the mainfest is being generated
+        for a subset of files in a Tale. Note that these records get added to the internal manifest
+        object.
+        """
+        for item_id in self.item_ids:
+            item = self.itemModel.load(item_id,
+                                       user=self.user,
+                                       level=AccessType.READ)
+            if item:
+                item_path = path_util.getResourcePath('item', item, user=self.user)
+                # Recreate the path
+                data_catalog_root = '/collection/' + CATALOG_NAME+'/'+CATALOG_NAME+'/'
+                workspaces_root = '/collection/' + WORKSPACE_NAME+'/'+WORKSPACE_NAME
+                # Check if the item belongs to workspace or external data
+                if item_path.startswith(workspaces_root):
+                    item_path = item_path.replace(workspaces_root, '')
+                    full_path = '../workspace' + clean_workspace_path(self.tale['_id'],
+                                                                      item_path)
+                    self.manifest['aggregates'].append({'uri': full_path})
+                    continue
+                elif item_path.startswith(data_catalog_root):
+                    item_path = item_path.replace(data_catalog_root, '')
+                    bundle = self.create_bundle('../data/' + item_path,
+                                                clean_workspace_path(self.tale['_id'],
+                                                                     item['name']))
+
+                    # Get the linkURL from the file object
+                    item_files = self.itemModel.fileList(item,
+                                                         user=self.user,
+                                                         data=False)
+                    for file_item in item_files:
+                        agg_record = self.create_aggregation_record(
+                            file_item[1]['linkUrl'],
+                            bundle,
+                            get_folder_identifier(item['folderId'],
+                                                  self.user))
+                        self.manifest['aggregates'].append(agg_record)
+                    self.datasets.add(item['folderId'])
```

Comment:
The assumption that item's parent folder will have all information about the original dataset is not true in general. It only works for packages registered as a folder with flat structure inside (DataONE/Dataverse), but will break for everything else (HTTP / Globus)

Reply:
In the case of Globus & HTTP, should I be looking inside the item's `meta` field?

Reply:
Not sure yet. Maybe it would be convenient to create API call for that, where you could consume item/folder and return provider specific dataset info. With Globus it's fairly easy, but I'm not sure what to do about plan http files.